Changes necessary to implement Scripting events for "game-select", "system-select" and "screeensaver-game-select"

EnsignRutherford opened the pull request:
Hi modified the core Scripting.h and Scripting.cpp at add an additional optional parameter to support triggering events and running script files that take "system name", "rom file name", "rom file game name" as parameters to their command-line.  I also modified the necessary files to trigger events for when the game selected via the UI changes, the system selected via the UI changes, or the screen saver switches videos or shows no video at all.  The changes allow easier interception of those events that I originally did to integrate the PixelCade LED Marquee with ES without having to use a custom build of ES to support it.  The logic that was modified I have moved and enhanced into the bash scripts that now get executed so developers can customize the behavior to their own liking.

For example, the default behavior of the Pixelcade is if you call it's RESTFUL API with the system, the rom name, and the game name of the rom it will try to display the marquee of the rom if it exists, else scroll the text name of the game.  I have modified this behavior in my script to see if the marquee file exists and if it doesn't exist display the marquee of the system.  There are probably other use cases or intercepting these events.

EnsignRutherford commented:
I had to add another instance of the "game-select" event to cover when the user has selected a system to startup with so it can display the proper game.

pjft commented:
Thanks for starting this off. I'll be honest, we'll need to assess how this goes and where/how to fit it in.

A comment to be mindful of: how can we reduce - or eliminate - the performance impact for the majority of users who will not really be using this? Or, alternatively, can you profile the performance and impact (see #716 ) of this change for us to be able to make an informed decision around this?

Thanks.

EnsignRutherford commented:
Hi,

 

I’m going on vacation so when I come back I’ll profile the changes to see how no shell scripts impact overall performance and will post my findings.

 

Without shell scripts it’s simply executing a check to see if files exist in the scripts folder.  That’s the performance hit.

 

Thanks.

 

 

From: pjft ***@***.***> 
Sent: Friday, May 21, 2021 12:26 PM
To: RetroPie/EmulationStation ***@***.***>
Cc: EnsignRutherford ***@***.***>; State change ***@***.***>
Subject: Re: [RetroPie/EmulationStation] Changes necessary to implement Scripting events for "game-select", "system-select" and "screeensaver-game-select" (#751)

 

Thanks for starting this off. I'll be honest, we'll need to assess how this goes and where/how to fit it in.

A comment to be mindful of: how can we reduce - or eliminate - the performance impact for the majority of users who will not really be using this? Or, alternatively, can you profile the performance and impact (see #716 <https://github.com/RetroPie/EmulationStation/pull/716>  ) of this change for us to be able to make an informed decision around this?

Thanks.

—
You are receiving this because you modified the open/close state.
Reply to this email directly, view it on GitHub <https://github.com/RetroPie/EmulationStation/pull/751#issuecomment-846082532> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AHPPESYUPLCBRBQOOEEEH2DTO2CSRANCNFSM45AYZEUQ> .  <https://github.com/notifications/beacon/AHPPES7J72JTJTCLNGBN6ZTTO2CSRA5CNFSM45AYZEU2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOGJXDDZA.gif>

EnsignRutherford commented:
The main additional code being called now is in the Scripting.cpp module in the fireEvent.  If the directory for the event does not exist the loop never gets executed...

 // check in exepath
        test = Utils::FileSystem::getExePath() + "/scripts/" + eventName;
        if(Utils::FileSystem::exists(test))
            scriptDirList.push_back(test);

        // check in homepath
        test = Utils::FileSystem::getHomePath() + "/.emulationstation/scripts/" + eventName;
        if(Utils::FileSystem::exists(test))
            scriptDirList.push_back(test);
        int ret = 0;
        // loop over found script paths per event and over scripts found in eventName folder.
        for(std::list<std::string>::const_iterator dirIt = scriptDirList.cbegin(); dirIt != scriptDirList.cend(); ++dirIt) {
            std::list<std::string> scripts = Utils::FileSystem::getDirContent(*dirIt);
            for (std::list<std::string>::const_iterator it = scripts.cbegin(); it != scripts.cend(); ++it) {

So that is the additional code that is being executed every time a game or system is selected.  With no scripts defined the performance impact should be minimal.  I've tested with and without scripts defined and don't see a difference in the user experience.  I can profile the code for profiling to verify.

EnsignRutherford commented:
I wrapped the code enhancements I added with ProfileBegin and ProfileEnd in all the files I touched and ran emulationstation with "--debug".  Upon exit I didn't get any profiling information.  What did I do wrong?

pjft commented:
If I recall correctly, you need to make/compile ES with Profiling enabled:

https://github.com/RetroPie/EmulationStation/commit/29223d9a2e9b6cee0b8487c04f0a7943540fbccc#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20a

EnsignRutherford commented:
Got it.  Here is the output:

Jun 10 16:25:56 lvl3:   Message                                        Calls      Total Time        Avg Time        Min Time        Max Time     Internal Total Time         Internal Avg Time
Jun 10 16:25:56 lvl3:   fireEvent for game-select                        173        0.027090        0.000157        0.000057        0.001142                0.027090                  0.000157
Jun 10 16:25:56 lvl3:   fireEvent for system-select                      123        0.010914        0.000089        0.000058        0.000545                0.010914                  0.000089
Jun 10 16:25:56 lvl3:   fireEvent for system-select - gotostart            1        0.000160        0.000160        0.000160        0.000160                0.000160                  0.000160

These timings are for when there are no scripts defined for those events.  This was generated by selecting every system once and then selecting one system and selecting every game in the system.  The internal average times are negligible.  :)

pjft commented:
Thanks for checking.

I'd very much appreciate @jrassa 's inputs here as he is a lot more familiar with the scripting engine code.

EnsignRutherford commented:
Ok.  I'm comfortable with the changes but I'd like a second opinion. :)

cmitu commented:
I have some comments after testing the changes in the PR:

* `system-select` is not consistently triggered, or maybe I haven't understood how it should work. Works fine if I change the system from the carousel, but it doesn't trigger when with fast system switching (i.e. left/right from an existing system). If I start the ES on a system, it gets triggered - so my understanding is that it should trigger also when entering a system (with 2nd parameter `requested` maybe ?).
* `game-select` triggers when shouldn't be necessary - for instance opening the menu(s) with Start/Select and then closing it triggers the event. The event should trigger only when the game is 1st selected.
* `screensaver-game-select` triggers even when it's not needed. I have the screensaver set to 'DIM' (so there's no video/image configured for it) and the log is spammed with `fireEvent: screensaver-game-select NULL NULL`. IMHO there shouldn't be any `NULL` sent to the script - the script should handle the arguments accordingly. The event should also not trigger when the screensaver is not configured for image/video display.

EnsignRutherford commented:
The main user story as to why I added those events was so that I can have scripts executed that sync the current system/game marquee on the Pixelcade with the one being selected/played/displayed

- The function that calls the fireEvent for "system-select" is not being called during fast system switching.  I prefer this behavior as switching the marquees that fast would cause a blur and definitely slowdown the system.

- The way I implemented it was to call fireEvent at the change event of the game or the system.  So "game-select" I see gets called twice: once for the current game and then a second time for the next game selected.  It also triggers when you enter or leave a system.  This is already pre-existing behavior in the code that calls the function where the fireEvent is now called.  

- I wanted to leave that event in with NULLS so the script has an opportunity to perform some action when the screensaver is not configured for image/video display.  In this instance, since the script called does not know what is on the screen it selects a random system/game marquee for the Pixelcade to display.

Here is the script that I have right now that is called for game-select, to give you an idea of how it's practically being used.
[https://github.com/EnsignRutherford/Pixelcade/blob/master/scripts/game-select/01-pixelcade.sh](url)

Currently the Pixelcade project compiles its own version of ES and has hardcoded calls to the REST services that the Pixelcade exposes.  I was hoping to make it more generic.

EnsignRutherford commented:
Should I create different events then?  I was using a parameter to also communicate to the script when the event was occuring, via "input" or "gotostart".  I'm trying to supply the script with sufficient information to decide what action to take.

The behavior of calling the code when the game is already selected is performed by the calling code.  Am I capturing that event at the wrong time?

I can remove that functionality and rely on screensaver start and stop events for DIM.  9059 is definitely excessive.

cmitu commented:
> The behavior of calling the code when the game is already selected is performed by the calling code. Am I capturing that event at the wrong time?

cmitu commented:
Sorry, accidentally hit close instead of comment.

> Should I create different events then? I was using a parameter to also communicate to the script when the event was occuring, via "input" or "gotostart". I'm trying to supply the script with sufficient information to decide what action to take.

If you're referring to the system select/browse, yes, it would be an option. Have different parameters for the 2 (or 3 situations) - `input` (or `selected`) when the system is focused on the carousel, `browse` for when the system is entered and maybe a 3rd one for when the system opened on first start (like it's right now).

> 
> The behavior of calling the code when the game is already selected is performed by the calling code. Am I capturing that event at the wrong time?

It may be a ES limitation, I haven't looked closer at the code. 
 
 
> I can remove that functionality and rely on screensaver start and stop events for DIM. 9059 is definitely excessive.

That would be preferable, yes.

EnsignRutherford commented:
Let me make some revisions, test and repost.  Great stuff!

EnsignRutherford commented:
Hi.  Is there any way to review and accept these changes?  Last update was June 17th.

obrie commented:
Just wanted to add that I have a similar use case for my own project and would love to help in any way I can (testing or otherwise).  Thanks @EnsignRutherford for working on this!

pjft commented:
Thanks. I'd like to get @jrassa 's review of this, if possible, as he is the most knowledgeable person about this part of ES.

EnsignRutherford commented:
Thanks.  I've put it through rigorous testing and my changes are solid.  I use those events to manage the display on a LED marquee in my bartop arcade build.

On August 16, 2021 6:52:40 AM EDT, Aaron Pfeifer ***@***.***> wrote:
>Just wanted to add that I have a similar use case for my own project and would love to help in any way I can (testing or otherwise).  Thanks @EnsignRutherford for working on this!
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/RetroPie/EmulationStation/pull/751#issuecomment-899414856
Russ

Sent from my Samsung Galaxy S20 Ultra

dale-personal commented:
I just want to say nice work.  I've just started looking into a marquee project myself and came across this project: https://github.com/alinke/PIXEL/tree/master/EmulationStation-kai but didn't like the idea of using a forked version of emulation station. This was the same solution i was thinking of, and would eliminate the need to fork (for pixelcade and other solutions I've seen).

EnsignRutherford commented:
> I just want to say nice work. I've just started looking into a marquee project myself and came across this project: https://github.com/alinke/PIXEL/tree/master/EmulationStation-kai but didn't like the idea of using a forked version of emulation station. This was the same solution i was thinking of, and would eliminate the need to fork (for pixelcade and other solutions I've seen).

Thanks.  Same thing here.  I didn't like using a separate fork of EmulationStation nor how it was implemented.  I have coded up all the scripts required to interphase with the Pixelcade.  They are on my github.

dale-personal commented:
> separate



> 
> 
> > I just want to say nice work. I've just started looking into a marquee project myself and came across this project: https://github.com/alinke/PIXEL/tree/master/EmulationStation-kai but didn't like the idea of using a forked version of emulation station. This was the same solution i was thinking of, and would eliminate the need to fork (for pixelcade and other solutions I've seen).
> 
> Thanks. Same thing here. I didn't like using a separate fork of EmulationStation nor how it was implemented. I have coded up all the scripts required to interphase with the Pixelcade. They are on my github.

These scripts are great - hope i can hit the ground running when my pixel board and panels arrive later this week. I'd like your advice / opinion on what I've been thinking.  I'll start by saying I'm running retropie on an x86 install of Ubuntu, and many of the pixelcade scripts I've seen have assumed a raspberry pi install and the existence of a /home/pi directory.  My concern is that the pixelcade APIs will assume this as well.  My approach was going to be to try and use this command line utility: https://ledpixelart.com/console/ in the hopes of gaining a little more control over where things are installed.

However, i get a java error when trying to run java -var pixelc.jar with any parameters.

Exception in thread "main" java.lang.NoClassDefFoundError: twitter4j/JSONException
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Class.java:398)
        at org.eclipse.jdt.internal.jarinjarloader.JarRsrcLoader.main(JarRsrcLoader.java:56)
Caused by: java.lang.ClassNotFoundException: twitter4j.JSONException
        at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)

I'm curious if this is something you've tried, and what you think about trying to use something like pixelc.jar along with placing marquee images in more standard emulationstation directories?  perhaps system marquees could go in /opt/retropie/configs/[system]/marquee.png for example, similar to how launching.png is done. or game marquees could go in /opt/retropie/configs/all/emulationstation/marquees/[system]/[rom].png

That was going to be my scripting approach, as it felt might make it more workable for install other than raspberrypi.

EnsignRutherford commented:
Going to take a look at this today.  I have another project I might be able to use this for.

dale-personal commented on es-app/src/views/gamelist/ISimpleGameListView.cpp:
```diff
@@ -160,6 +161,15 @@ bool ISimpleGameListView::input(InputConfig* config, Input input)
+	FileData* cursor = getCursor();
+	SystemData* system = this->mRoot->getSystem();
+    	if (system != NULL) {
+            Scripting::fireEvent("game-select", system->getName(), cursor->getFileName(), cursor->getName(), "input");
```
Consider using cursor->getSystemName() here instead of system->getName(). for game lists like favorites to receive the actual system name instead of "favorites"

EnsignRutherford replied:
Actually that is the behavior I was looking for.  For example, if someone is in a custom collection e.g "Aliens" and they have a ROM that doesn't have artwork for the pixelcade specifically for it, this code will tell it to display the "Aliens" system artwork.  The idea being the default shows where you are, not system the game is from.

dale-personal reviewed (commented):
custom game list feedback from testing.

EnsignRutherford commented:
Hi,

This has been opened since May 17th.  I'm using my own custom build without issue thanks to all the input.  How does this get resolved, if at all?

Rutherford

alinke commented:
joining in on the conversation, I'm the creator of the Pixelcade marquees and would be really nice to switch over to the ES event engine that this PR would enable vs. the current fork with hard coded REST calls we use for the Pixelcade integration. And note this is working on Batocera and EmuELEC now, example here https://pixelcade.org/emuelec/ . Would be much appreciated if this PR could be reviewed, thanks.

dale-personal commented:
Quick Update. I have been running these changes for the last 6 months, and can vouch for the stability and performance. My testing has been on Ubuntu 21.10 on AMD Ryzen 2500u hardware.

tomaz82 commented on es-app/src/SystemScreenSaver.cpp:
```diff
@@ -11,7 +11,9 @@
 #include "FileData.h"
 #include "FileFilterIndex.h"
 #include "Log.h"
+#include "utils/ProfilingUtil.h"
```
This include could be removed again

tomaz82 commented:
If that ProfilingUtil include is removed then I'm ok with this being merged.

EnsignRutherford commented:
I removed the include and updated git appropriately.

cmitu commented:
I've re-checked the changes and my only objection would be the lack of a `system-select` event when changing systems via fast switching - see my previous comment on https://github.com/RetroPie/EmulationStation/pull/751#issuecomment-860842883.

EnsignRutherford commented:
I help off capturing that 'event' because at least for the use case of the PixelCade I wouldn't want it to 'flutter'.  Let me experiment with that and see.

EnsignRutherford commented:
I also added the name of the rom in the game-start event per a request from Al Linke at PixelCade.

tomaz82 commented:
I've read through the code and the comments.
It just needs a rebase and then I'm happy to merge this.

EnsignRutherford commented:
Sometimes I hate git.... how do I do that in an already existing branch?  I see there are conflicts with the SystemScreenSaver.cpp.  Can I just merge the changes manually?

tomaz82 commented:
I can try to fix the rebasing for you

tomaz82 commented:
I think I finally managed to piece it together :P
Thanks!
